### PR TITLE
NAU7802 implementation and supporting refactor

### PR DESCRIPTION
**Pull request by fivesixzero**
Refactor to support the NAU7802 ADC for weight measurements in addition to the HX711. More details can be found in issue #32.

### Goals

1. Refactor backend to allow for a common `Scale` interface
2. Create HX711 driver or wrapper an for existing driver to use the new interface
3. Create NAU7802 driver or wrapper for an existing driver to use the new interface
5. Add UI elements to settings interface to allow for selecting which driver to use
6. Add UI elements to settings interface to allow for configuring lower-level per-driver settings, like HX711 clock/data pin and NAU7802 I2C bus ID
7. Implement settings migration to allow for automatic automatic setup of legacy HX711-based installation and [settings versioning](https://docs.octoprint.org/en/master/plugins/mixins.html#settingsplugin) to allow for more easily updating configuration between versions moving forward

### Related Open Issues

This will address issue #32 and may also fully or partially address #28, #26, and #18.

Moving to use of a common "scale" interface and settings versioning should also lay the groundwork to allow for addressing #31, #22, and #17.

### CODE REVIEW

**Comment by fivesixzero** [comment]
The first commit is rough but it should take care of a few of the goals listed above.

1. A common `Scale` interface was added and the backend was refactored to allow for this new interface.
    * Some basic changes were made the frontend as well to allow for tare/calibrate operations using this new interface.
2. A wrapper implementing the `Scale` interface was created for the [`hx711`](https://pypi.org/project/hx711/) driver, which is a pretty robust and easy to implement driver for the HX711 on a Raspberry Pi.
    * This was tested with a common HX711 breakout with a simple filament scale configuration.
3. A wrapper implementing the `Scale` interface was created for the [`PyNAU7802`](https://pypi.org/project/PyNAU7802/) driver)
    * This was tested with both Adafruit and Sparkfun's NAU7802 breakouts with a simple filament scale configuration.

I ran out of time to start in earnest on the frontend changes, settings migration, and basic unit/integration testing, so those remain left to implement before this PR can be considered for proper review/merge.

File: filament_scale_enhanced/__init__.py [hunk 1]
```diff
@@ -46,19 +44,63 @@ def get_assets():
...
+            from .scale.hx711 import HX711
+
+            clock_pin = self._settings.get(["hx_clockpin"])
+            data_pin = self._settings.get(["hx_datapin"])
+
+            self._logger.debug("Initializing HX711, clock_pin: [{}], data_pin: [{}]".format(clock_pin, data_pin))
+            self.scale = HX711(offset, cal_factor, clock_pin, data_pin)
+            self._logger.debug("Initialized HX711, is_ready: [{}]".format(self.scale.is_ready()))
+        else:
+            self.scale = None
```
Comment (techman83) [inline filament_scale_enhanced/__init__.py:75]: It'll save some boiler plate and give room for different backends in the future (maybe say an external arduino), if we `from . import scale` at the top of the file, then use getattr and pass the whole config payload into the library. Let each implementation handle parsing it during instantiation. Along with the logging statements. Potentially a try/except for if the backend configured doesn't exist

```suggestion
    self.scale = getattr(scale, chip)(settings=self._settings)
```

File: filament_scale_enhanced/__init__.py [hunk 2]
```diff
@@ -46,19 +44,63 @@ def get_assets():
...
+        self._logger.debug("weight: [{}]".format(v))
+
+        self._settings.set(["lastknownweight"], v)
+        self._settings.save()
+
         self._plugin_manager.send_plugin_message(self._identifier, v)
-        self.hx.power_down()
+
+        if chip == "hx711":
+            self.scale.disable()
```
Comment (techman83) [inline filament_scale_enhanced/__init__.py:102]: I can't recall why it's necessary to power up/down here for the hx11. But maybe implementing a `initial_check_weight` method, that by default calls `get_weight` and is overriden by the hx711 chip might be a touch cleaner

Something like this in the hx11 class
```python
def initial_check_weight(self) -> float:
    self.enable()
    weight = self.get_weight()
    self.disable()
    return weight
```

**Review by techman83 — COMMENTED** [review]
Really cool! I'll ponder the testing side of things. I want to test the various migration paths in particular.

File: filament_scale_enhanced/__init__.py [hunk 3]
```diff
@@ -30,12 +24,16 @@ def get_template_configs():
     @staticmethod
     def get_settings_defaults():
         return dict(
-            tare=8430152,
-            reference_unit=-411,
+            # chips=["hx711", "nau7802", "none"],
+            chip="nau7802",
```
[the remark this answers is not in the document]
Comment (fivesixzero) [reply filament_scale_enhanced/__init__.py:28]: Definitely agree there. That default was a debug change that I forgot to change out before commit. 🤦 

- [ ] Assure that default `chip` is `hx711`, not `nau7802`

File: filament_scale_enhanced/__init__.py (hunk 1 above)
Comment (fivesixzero) [reply filament_scale_enhanced/__init__.py:75]: That's a great idea and I'm all for simplifying this code.

- [ ] Move per-device instantiation details to `scale` library
- [ ] Implement config parsing in `scale` library to determine which subclass to init and handle per-subclass variables
- [ ] Use `getarr()` to init `self.scale` with the active config using the above implementation during `on_startup()`

File: filament_scale_enhanced/__init__.py (hunk 2 above)
Comment (fivesixzero) [reply filament_scale_enhanced/__init__.py:102]: This was one of the first parts of the code I dug into and this is an example of just trying to keep things as identical as possible to the original implementation before making more changes.

When I was working on the [CircuitPython HX711 GPIO bitbang driver](https://github.com/fivesixzero/CircuitPython_HX711/blob/main/hx711/hx711_gpio.py) I definitely observed some oddities around 'enable' vs 'disable' state management. In that driver's case I basically just abandoned all hope of tracking a 'powered up' vs 'powered down' state and pared the implementation down to the barest of basics, the raw ADC count read. That has its own downside - leaving the clock line low means the first read-in could provide a stale reading - but its easy enough to just read a few dozen counts, throw away the first read (or just remove outliers) before averaging.

Even with that knowledge though there's always going to be a risk of severe timing issues causing odd behavior in general, which is kinda unavoidable when bitbanging from high in the stack like this. Even with the much more lightweight CircuitPython, platforms like `nrf` and `esp32` (which do BLE/WiFi management in background threads) can see some [weird timing issues](https://github.com/fivesixzero/CircuitPython_HX711/issues/2) pop up that can totally wreck the driver's reliability. So this kind of mitigation doesn't surprise me or seem out of place, even if I'm not sure precisely why it was included.

Anyway, long story short, I think this "state toggle" should probably stick around (for now, at least). And I agree that it probably belongs in the `scale.hx711` wrapper rather than here in the main plugin code. :)

**Comment by fivesixzero** [comment]
Finally got some time set aside to get my dev environment reconfigured and dig into this again. Should have a few solid, multi-hour chunks of time to dive in over the next few days.

Right off the bat I found a few embarassing typos that broke things, which I fixed in e6d858d. These slipped in because my original setup involved some copy/paste from my Pi devbox to my main dev PC (gross!). Now I'm just doing git commits directly from the devbox with my dev PC connecting via the [VSCode `sshfs` plugin](https://marketplace.visualstudio.com/items?itemName=Kelvin.vscode-sshfs) for easier editing. No more copy/paste errors now. :)

On the dev system I've got two filament scales attached, one via HX711 on the usual pin 20/21 combo and the other via a NAU7802 on I2C. This way I can easily switch between the two by changing the setting in the Octoprint config.

Not sure what the order of operations is going to be but I've got a few chunks of to-do items.

1. Refactor `scale` lib to take Octoprint settings as input and handle per-device instantiation
    * As a side goal, I'd like to make it easier to re-init the scale without restarting the whole app, which would allow for proper settings changes.
2. Rest API auth
    * If there's a REST API it should require and use an API token, at the very least
3. Refactor JS and templates
    * It looks like there have been some improvements to Octoprint's plugin API since the original JS/templates were created so it could use a refresh to make it easier to work with
4. Start investigating settings versioning and migration
    * Octoprint looks like it has [decent functionality available](https://docs.octoprint.org/en/master/plugins/mixins.html#octoprint.plugin.SettingsPlugin.get_settings_version) that should help to organize this but I've only barely scratched the surface there
5. Implement proper unit tests for the `scale` library
    * I'm still new to Pythonic unit testing but I'm eager to learn, so this may actually be a higher priority since I'm a bit more comfortable working with at least partially test-driven dev cycles

**Comment by fivesixzero** [comment]
Latest commit tackles a lot more than I like to bundle into a single commit but most of it had to land at the same time, so I'll let it slide. 😎 

- Changed `chip` setting to `scale_type` to reflect possible future scale options
    - Future `Scale` implementations could pull scale data from an external REST API, an MQTT or other message queue, platform-specific GPIO/SPI/I2C connections, or any other source
- Refactored `scale` module
    - New `get_scale(settings: Dict)` simplifies scale init
    - Added `deinit()` method to `Scale` interface for releasing resources like GPIOs, I2C busses, etc
    - Changed `hx711` to `rpi_hx711` to reflect the `HX711` driver's platform requirements
    - Moved pre- and post-read-in HX711 enable/disable into the `rpi_hx711` driver wrapper
    - Considered changing `nau7802` as well but this driver's `smbus2` implementation may work fine on a variety of platforms
- Moved `self.scale` and UI-updating `RepeatingTimer` init to helper methods and added deinit helpers as well
- Added new API endpoints in `__init__.on_api_get()`
    - `spool_weight` allows for get/set of spool weight variable
    - `scale_type` allows for get/set of scale_type setting, including deinit/reinit of `self.scale` if a change is made
- Added empty stubs for `get_settings_version` and `on_settings_migrate` to help with future settings migration implementation

This iteration doesn't use `getattr()`, but it does return a scale given a settings dict. Between that and breaking out the two main operations - scale init and timer init - into helper methods the `on_startup()` method is heavily simplified, which is good. Let me know if this looks good or if another implementation would be better, I'm still learning the Pythonic way and I'm flexible. :)

Also rearranged a lot of code to group similar things together and satisfy linters. My first commit was a total mess. :joy:

These are the linters/formatters I tested with and their basic configs, if different from the project's config files:

- `flake8`
    - No changes
- `pylint`
    - `max-line-length=140` to match project's `flake8` config
    - Added `init-hook` for `pylint-venv` to handle OctoPrint dev environment venv imports
- `mypy`
    - `mypy --disallow-untyped-defs --disallow-untyped-calls --disallow-incomplete-defs  --check-untyped-defs --disallow-untyped-decorators  ./filament_scale_enhanced/scale/`
- `black` (not in project, yet, but handy)
    - `--line-length 140` to match project's `flake8` config

If these linter settings are good then I can enshrine most (or all) of them in their respective configs and add them to the GitHub `test.yml` workflow. 

I've used [`pre-commit`](https://pre-commit.com/) on a few other projects, which can help with running linters/tests in a consistent way during ongoing development. I may look at implementing that here if you're cool with it.

Next stuff...

Didn't dive into REST API Auth, JS/templates, settings versioning, or unit test implementation in these changes but I spent a bit of time researching each as I worked through things. Of those, the JS/template work is probably the foggiest in my head at the moment. For the others I have at least some basic implementation ideas/examples in mind.

Let me know if you've got any feedback on the latest state of the code, or if I've missed implementing any changes suggested on the first commit. If all goes well I should have another few chunks of time like this in the coming days.

File: filament_scale_enhanced/__init__.py [hunk 4]
```diff
@@ -70,18 +98,210 @@ def get_update_information(self):
...
+    # pylint: disable=too-many-return-statements,too-many-branches,too-many-statements,broad-except
+    def on_api_get(self, request) -> flask.Response:
+        if time.time() - self.last_api_time < 0.1:
+            return flask.make_response("Slow down!", 200)
+
+        command = request.args.get("command")
+        value = request.args.get("value")
+
+        if command == "tare":
+            self.last_api_time = time.time()
```
Comment (techman83) [inline filament_scale_enhanced/__init__.py:204]: A method I often use here is
```python
getattr(self, command)()
```

This makes adding new functionality just adding a method, keeps the entry function smaller, and makes unittesting easier.

As it all involves hardware, unit testing is probably hard. But we might be able to write a mock scale class. But I wouldn't consider that level required to merge.

File: filament_scale_enhanced/scale/__init__.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+def get_scale(settings: Dict) -> Scale:
+    """Returns a fully configured Scale based on the settings provided"""
+
+    scale_type = settings["scale_type"]
+
+    if scale_type not in SCALE_TYPES:
+        raise ValueError(f"Invalid scale_type: {scale_type}")
+
+    if scale_type == "rpi_hx711":
+        from .rpi_hx711 import RPI_HX711
```
Comment (techman83) [inline filament_scale_enhanced/scale/__init__.py:17]: I think I mentioned an importlib method previously. This keeps the logic out of the main entry point which is good.

**Review by techman83 — COMMENTED** [review]
I need to spend a good chunk of time reading this, but I think it's great! I'll try and keep up to date and appreciate the effort (I've lacked spare time lately!)

**Comment by fivesixzero** [comment]
Oof, hope you're feeling better! I ended up getting sidetracked into other tasks over the last few weeks but I'm hoping to have more time to dive back in next week.

**Comment by techman83** [comment]
> Oof, hope you're feeling better! I ended up getting sidetracked into other tasks over the last few weeks but I'm hoping to have more time to dive back in next week.

Turns out Covid doesn't mix well with ADHD! Took months for the brain fog to lift, and I'm just starting to get enough energy back to keep up with work. But I have been 3D printing again! So I'm going to try and get through this, and try and come up with ways to make it easier to approve/merge/release changes.
